ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/flat/RewardTraceGenerator.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.flat;
+
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.Trace;
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.Wei;
+import org.hyperledger.besu.ethereum.mainnet.MiningBeneficiaryCalculator;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSpec;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+
+public class RewardTraceGenerator {
+
+  private static final String REWARD_LABEL = "reward";
+  private static final String BLOCK_LABEL = "block";
+  private static final String UNCLE_LABEL = "uncle";
+
+  /**
+   * Generates a stream of reward {@link Trace} from the passed {@link Block} data.
+   *
+   * @param protocolSchedule the {@link ProtocolSchedule} to use
+   * @param block the current {@link Block} to use
+   * @return a stream of generated reward traces {@link Trace}
+   */
+  public static Stream<Trace> generateFromBlock(
+      final ProtocolSchedule<?> protocolSchedule, final Block block) {
+
+    final List<Trace> flatTraces = new ArrayList<>();
+
+    final BlockHeader blockHeader = block.getHeader();
+    final ProtocolSpec<?> protocolSpec = protocolSchedule.getByBlockNumber(blockHeader.getNumber());
+    final MiningBeneficiaryCalculator miningBeneficiaryCalculator =
+        protocolSpec.getMiningBeneficiaryCalculator();
+
+    final AtomicReference<Wei> totalBlockReward =
+        new AtomicReference<>(protocolSpec.getBlockReward());
+
+    // add uncle reward traces
+    block
+        .getBody()
+        .getOmmers()
+        .forEach(
+            ommerBlockHeader -> {
+              final long distance = blockHeader.getNumber() - ommerBlockHeader.getNumber();
+              final Wei blockReward = protocolSpec.getBlockReward();
+              final Wei ommerReward =
+                  blockReward.subtract(blockReward.multiply(distance).divide(8));
```

Comment:
Is 8 here a magic number?

Reply:
This is a magic number, along with 32, and *yikes* it's everywhere.  We need to factor it out but that is out of scope for this PR, I opened #452 and added a good first issue tag.

Reply:
I just made a modification that uses the BlockProcessor and I added tests to verify that the RewardTracegenerator is using the processor to calculate the rewards